ddtrace/profiling/collector/_memalloc_heap.cpp

```diff
@@ -107,226 +149,268 @@ heap_tracker_next_sample_size(uint32_t sample_size)
     return (uint32_t)(log_val * (-log(2) * (sample_size + 1)));
 }
 
-static void
-heap_tracker_init(heap_tracker_t* heap_tracker)
+// Method implementations
+heap_tracker_t::heap_tracker_t(uint32_t sample_size_val)
+  : sample_size(sample_size_val)
+  , current_sample_size(heap_tracker_next_sample_size(sample_size_val))
+  , allocated_memory(0)
+  , frozen(false)
 {
-    heap_tracker->allocs_m = memalloc_heap_map_new();
-    heap_tracker->freezer.allocs_m = memalloc_heap_map_new();
-    ptr_array_init(&heap_tracker->freezer.frees);
-    traceback_array_init(&heap_tracker->unreported_samples);
-    heap_tracker->allocated_memory = 0;
-    heap_tracker->frozen = false;
-    heap_tracker->sample_size = 0;
-    heap_tracker->current_sample_size = 0;
-    memalloc_gil_debug_check_init(&heap_tracker->gil_guard);
+    // gil_guard, allocs_m, and freezer_allocs_m are initialized by their constructors
 }
 
-static void
-heap_tracker_wipe(heap_tracker_t* heap_tracker)
+heap_tracker_t::~heap_tracker_t()
 {
-    memalloc_heap_map_delete(heap_tracker->allocs_m);
-    memalloc_heap_map_delete(heap_tracker->freezer.allocs_m);
-    ptr_array_wipe(&heap_tracker->freezer.frees);
-    traceback_array_wipe(&heap_tracker->unreported_samples);
+    // std::vector and memalloc_heap_map destructors handle cleanup automatically
 }
 
-static void
-heap_tracker_freeze(heap_tracker_t* heap_tracker)
+void
+heap_tracker_t::freeze()
 {
-    MEMALLOC_GIL_DEBUG_CHECK_ACQUIRE(&heap_tracker->gil_guard);
-    assert(!heap_tracker->frozen);
-    heap_tracker->frozen = true;
-    MEMALLOC_GIL_DEBUG_CHECK_RELEASE(&heap_tracker->gil_guard);
+    MEMALLOC_GIL_DEBUG_CHECK_ACQUIRE(&gil_guard);
+    assert(!frozen);
+    frozen = true;
+    MEMALLOC_GIL_DEBUG_CHECK_RELEASE(&gil_guard);
 }
 
-/* Un-freeze the profiler, and return any samples we weren't able to remove while
- * the profiler was frozen. This function modifies the profiler state, so it must
- * be called with the GIL held and must not call any C Python APIS. */
-static traceback_t**
-heap_tracker_thaw_no_cpython(heap_tracker_t* heap_tracker, size_t* n_to_free)
+std::vector<traceback_t*>
+heap_tracker_t::thaw_no_cpython()
 {
-    MEMALLOC_GIL_DEBUG_CHECK_ACQUIRE(&heap_tracker->gil_guard);
-    assert(heap_tracker->frozen);
+    MEMALLOC_GIL_DEBUG_CHECK_ACQUIRE(&gil_guard);
```

Comment:
We could perhaps leverage RAII a bit more here and do something like https://github.com/DataDog/dd-trace-py/blob/27395c53d6f696e3a2720179dc4d2e325f624272/ddtrace/internal/_threads.cpp#L18-L19